csharp/ql/src/Likely Bugs/ThreadUnsafeICryptoTransform.ql

```diff
@@ -0,0 +1,93 @@
+ *              Using this an instance of this class in concurrent threads is dangerous as it may not only result in an error, 
+ *              but under some circumstances may also result in incorrect results.
+ * @kind problem
+ * @problem.severity warning
+ * @precision medium
+ * @id cs/thread-unsafe-icryptotransform-field-in-class
+ * @tags concurrency
+ *       security
+ *       external/cwe/cwe-362
+ */
+
+import csharp
+
+class ICryptoTransform extends Class {
+  ICryptoTransform() {
+    this.getABaseType*().hasQualifiedName("System.Security.Cryptography", "ICryptoTransform")
+  }
+}
+
+predicate usesICryptoTransformType( Type t ) {
+  exists(  ICryptoTransform ict |
+    ict = t
+    or usesICryptoTransformType( t.getAChild*() )
+  )
+}
+
+predicate hasICryptoTransformMember( Class c) {
+  exists( Field f |
+    f = c.getAMember*()
+    and ( 
+      exists( ICryptoTransform ict | ict = f.getType() )
+      or hasICryptoTransformMember(f.getType())
+      or usesICryptoTransformType(f.getType())
+    )
+  )
+}
+
+predicate hasICryptoTransformStaticMemberNested( Class c ) {
+  exists( Field f |
+    f = c.getAMember() |
+    hasICryptoTransformStaticMemberNested( f.getType() )
+    or (
+      f.isStatic() and hasICryptoTransformMember(f.getType())
+      and not exists( Attribute a
+        | a = f.getAnAttribute() | 
+        a.getType().getQualifiedName() = "System.ThreadStaticAttribute"
+      )
+    )
+  )
+}
+
+predicate hasICryptoTransformStaticMember( Class c, string msg) {
+  exists( Field f |
+    f = c.getAMember*()
+    and f.isStatic()
+    and not exists( Attribute a
+      | a = f.getAnAttribute() 
+      and a.getType().getQualifiedName() = "System.ThreadStaticAttribute"
+    )
+    and (
+      exists( ICryptoTransform ict | 
+        ict = f.getType() 
+        and msg = "Static field " + f + " of type " + f.getType() + ", implements 'System.Security.Cryptography.ICryptoTransform', but it does not have an attribute [ThreadStatic]. The usage of this class is unsafe for concurrent threads."
+      )
+      or 
+      ( 
+        not exists( ICryptoTransform ict | ict = f.getType() ) // Avoid dup messages
+        and exists( Type t | t = f.getType() |  
+          usesICryptoTransformType(t)
+          and msg = "Static field " + f + " of type " + f.getType() + " makes usage of 'System.Security.Cryptography.ICryptoTransform', but it does not have an attribute [ThreadStatic]. The usage of this class is unsafe for concurrent threads."
+        )
+      )
+    )
+  )
+  or
+  exists( Field f |
+    f = c.getAMember*()
+    and not f.isStatic()
+    and ( hasICryptoTransformStaticMember( f.getType(), _ )
+      and msg = "Non-static field " + f + " of type " + f.getType() + " internally makes use of an static object that implements 'System.Security.Cryptography.ICryptoTransform'. This causes that usage of this class member is unsafe for concurrent threads."
```

Comment:
Perhaps this case should be limited to situations where the underlying static field is not in source code? Otherwise I would expect the underlying field to have already been flagged by one of the cases above.

Reply:
Also, an static -> a static

Reply:
If I remove this one, as you mentioned the underlying field class would have been flagged anyway.
I was looking for corner cases, but perhaps it is cleaner without it.
Giving it a try without this case.

Reply:
Please let me know what do you think about this one, or if I should add something that looks for cases where the underlying field is not in source code (the culprit class should still be flagged either way).

Reply:
I will let you decide, either remove this case, or restrict it to situations where the underlying static field is not in source code.

Reply:
I removed the case.
BTW. I am working on a separate QL that will try to find if any Lambda is capturing a local variable that is of type ICryptoTransform (or has a member of this type). The new QL should detect cases that this one cannot such as the stackoverflow example & complement each other. I will probably have it ready next week.